source/common/grpc/json_transcoder_filter.cc

```diff
@@ -0,0 +1,398 @@
+                                method_descriptor->client_streaming(), true)};
+
+  const auto response_type_url =
+      TYPE_URL_PREFIX + "/" + method_descriptor->output_type()->full_name();
+  std::unique_ptr<ResponseToJsonTranslator> response_translator{
+      new ResponseToJsonTranslator(type_helper_->Resolver(), response_type_url,
+                                   method_descriptor->server_streaming(), &response_input)};
+
+  transcoder.reset(
+      new TranscoderImpl(std::move(request_translator), std::move(response_translator)));
+  return Status::OK;
+}
+
+Status JsonTranscoderConfig::methodToRequestInfo(const google::protobuf::MethodDescriptor* method,
+                                                 google::grpc::transcoding::RequestInfo* info) {
+  auto request_type_url = TYPE_URL_PREFIX + "/" + method->input_type()->full_name();
+  info->message_type = type_helper_->Info()->GetTypeByTypeUrl(request_type_url);
+  if (info->message_type == nullptr) {
+    ENVOY_LOG(debug, "Cannot resolve input-type: {}", method->input_type()->full_name());
+    return Status(Code::NOT_FOUND, "Could not resolve type: " + method->input_type()->full_name());
+  }
+
+  return Status::OK;
+}
+
+JsonTranscoderFilter::JsonTranscoderFilter(JsonTranscoderConfig& config) : config_(config) {}
+
+Http::FilterHeadersStatus JsonTranscoderFilter::decodeHeaders(Http::HeaderMap& headers,
+                                                              bool end_stream) {
+  const auto status =
+      config_.createTranscoder(headers, request_in_, response_in_, transcoder_, method_);
+
+  if (!status.ok()) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  headers.removeContentLength();
+  headers.insertContentType().value(Http::Headers::get().ContentTypeValues.Grpc);
+  headers.insertPath().value("/" + method_->service()->full_name() + "/" + method_->name());
+
+  headers.insertMethod().value(Http::Headers::get().MethodValues.Post);
+
+  headers.insertTE().value(Http::Headers::get().TEValues.Trailers);
+
+  if (end_stream) {
+    request_in_.finish();
+
+    const auto& request_status = transcoder_->RequestStatus();
+    if (!request_status.ok()) {
+      ENVOY_LOG(debug, "Transcoding request error " + request_status.ToString());
+      error_ = true;
+      Http::Utility::sendLocalReply(*decoder_callbacks_, Http::Code::BadRequest,
+                                    request_status.error_message().ToString());
+
+      return Http::FilterHeadersStatus::StopIteration;
+    }
+
+    Buffer::OwnedImpl data;
+    readToBuffer(*transcoder_->RequestOutput(), data);
+
+    if (data.length()) {
+      decoder_callbacks_->addDecodedData(data);
+    }
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus JsonTranscoderFilter::decodeData(Buffer::Instance& data, bool end_stream) {
+  if (error_) {
+    return Http::FilterDataStatus::StopIterationNoBuffer;
```

Comment:
I don't think this can happen if you send a local complete reply. I would ASSERT(!error_)